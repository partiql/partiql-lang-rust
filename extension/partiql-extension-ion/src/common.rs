/// The encoding to use when decoding/encoding Ion to/from PartiQL [`partiql_value::Value`]
#[derive(Copy, Clone, Eq, PartialEq)]
pub enum Encoding {
    /// 'Unlifted'/'Unlowered' Ion to/from PartiQL [`partiql_value::Value`]. [`partiql_value::Value`]s that do not have a direct
    /// Ion analog will result in an error (e.g. PartiQL [`partiql_value::Value`] has a 'bag' type, but Ion does not,
    /// so attempting to encode a 'bag' results in an error).
    Ion,
    /// PartiQL encoded into Ion as supported by `partiql-tests` and by other implementations.
    /// (e.g., a PartiQL bag is encoded as an Ion list annotated with "$bag").
    PartiqlEncodedAsIon,
}

pub(crate) const BAG_ANNOT: &str = "$bag";
pub(crate) const TIME_ANNOT: &str = "$time";
pub(crate) const DATE_ANNOT: &str = "$date";
pub(crate) const MISSING_ANNOT: &str = "$missing";

pub(crate) const TIME_PART_HOUR_KEY: &str = "hour";
pub(crate) const TIME_PART_MINUTE_KEY: &str = "minute";
pub(crate) const TIME_PART_SECOND_KEY: &str = "second";
pub(crate) const TIME_PART_TZ_HOUR_KEY: &str = "timezone_hour";
pub(crate) const TIME_PART_TZ_MINUTE_KEY: &str = "timezone_minute";

pub(crate) const RE_SET_TIME_PARTS: [&str; 5] = [
    "^hour$",
    "^minute$",
    "^second$",
    "^timezone_hour$",
    "^timezone_minute$",
];
pub(crate) const TIME_PARTS_HOUR: usize = 0;
pub(crate) const TIME_PARTS_MINUTE: usize = 1;
pub(crate) const TIME_PARTS_SECOND: usize = 2;
pub(crate) const TIME_PARTS_TZ_HOUR: usize = 3;
pub(crate) const TIME_PARTS_TZ_MINUTE: usize = 4;
