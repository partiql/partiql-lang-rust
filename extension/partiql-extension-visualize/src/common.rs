pub trait ToDotGraph<T> {
    fn to_graph(self, data: &T) -> String;
}

#[cfg(feature = "visualize-dot")]
pub(crate) const FG_COLOR: &str = "\"#839496\"";
