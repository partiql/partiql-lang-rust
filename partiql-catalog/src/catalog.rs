use crate::call_defs::ScalarFnCallSpecs;
use crate::scalar_fn::ScalarFunction;
use crate::table_fn::TableFunction;
use partiql_common::catalog::{CatalogId, EntryId, ObjectId};
use partiql_types::PartiqlShape;
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::atomic::{AtomicU64, Ordering};
use thiserror::Error;
use unicase::UniCase;

/// Contains the errors that occur during Catalog related operations
#[derive(Error, Debug, Clone, PartialEq)]
#[error("Catalog error: encountered errors")]
pub struct CatalogError {
    pub errors: Vec<CatalogErrorKind>,
}

impl CatalogError {
    #[must_use]
    pub fn new(errors: Vec<CatalogErrorKind>) -> Self {
        CatalogError { errors }
    }
}

/// Catalog Error kind
///
/// ### Notes
/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
#[derive(Error, Debug, Clone, PartialEq)]
#[non_exhaustive]
pub enum CatalogErrorKind {
    /// Entry exists error.
    #[error("Catalog error: entry already exists for `{0}`")]
    EntryExists(String),

    /// Entry error.
    #[error("Catalog error: `{0}`")]
    EntryError(String),

    /// Any other catalog error.
    #[error("Catalog error: unknown error")]
    Unknown,
}

pub trait Catalog: Debug {
    fn add_table_function(&mut self, info: TableFunction) -> Result<ObjectId, CatalogError>;
    fn add_scalar_function(&mut self, info: ScalarFunction) -> Result<ObjectId, CatalogError>;

    fn add_type_entry(&mut self, entry: TypeEnvEntry<'_>) -> Result<ObjectId, CatalogError>;

    fn get_function(&self, name: &str) -> Option<FunctionEntry<'_>>;

    fn get_function_by_id(&self, id: ObjectId) -> Option<FunctionEntry<'_>>;

    fn resolve_type(&self, name: &str) -> Option<TypeEntry>;
}

#[derive(Debug)]
pub struct TypeEnvEntry<'a> {
    name: UniCase<String>,
    aliases: Vec<&'a str>,
    ty: PartiqlShape,
}

impl<'a> TypeEnvEntry<'a> {
    #[must_use]
    pub fn new(name: &str, aliases: &[&'a str], ty: PartiqlShape) -> Self {
        TypeEnvEntry {
            name: UniCase::from(name.to_string()),
            aliases: aliases.to_vec(),
            ty,
        }
    }
}

#[derive(Debug, Clone)]
pub struct TypeEntry {
    id: ObjectId,
    ty: PartiqlShape,
}

impl TypeEntry {
    #[must_use]
    pub fn id(&self) -> &ObjectId {
        &self.id
    }

    #[must_use]
    pub fn ty(&self) -> &PartiqlShape {
        &self.ty
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct FunctionEntry<'a> {
    id: ObjectId,
    function: &'a FunctionEntryFunction,
}

#[derive(Debug)]
pub enum FunctionEntryFunction {
    Table(TableFunction),
    Scalar(ScalarFnCallSpecs),
    Aggregate(),
}

impl<'a> FunctionEntry<'a> {
    pub fn id(&self) -> &ObjectId {
        &self.id
    }

    #[must_use]
    pub fn entry(&'a self) -> &'a FunctionEntryFunction {
        self.function
    }
}

#[derive(Debug)]
pub struct PartiqlCatalog {
    functions: CatalogEntrySet<FunctionEntryFunction>,
    types: CatalogEntrySet<PartiqlShape>,
    id: CatalogId,
}

impl Default for PartiqlCatalog {
    fn default() -> Self {
        PartiqlCatalog {
            functions: Default::default(),
            types: Default::default(),
            id: 1.into(),
        }
    }
}

impl PartiqlCatalog {}

impl Catalog for PartiqlCatalog {
    fn add_table_function(&mut self, info: TableFunction) -> Result<ObjectId, CatalogError> {
        let call_def = info.call_def();
        let names = call_def.names.clone();
        if let Some((name, aliases)) = names.split_first() {
            let eid = self
                .functions
                .add(name, aliases, FunctionEntryFunction::Table(info))?;
            Ok((self.id, eid).into())
        } else {
            Err(CatalogError::new(vec![CatalogErrorKind::EntryError(
                "Function definition has no name".into(),
            )]))
        }
    }

    fn add_scalar_function(&mut self, info: ScalarFunction) -> Result<ObjectId, CatalogError> {
        let id = self.id;
        let call_def = info.into_call_def();
        let names = call_def.names;
        if let Some((name, aliases)) = names.split_first() {
            self.functions
                .add(
                    name,
                    aliases,
                    FunctionEntryFunction::Scalar(call_def.overloads),
                )
                .map(|eid| ObjectId::new(id, eid))
        } else {
            Err(CatalogError::new(vec![CatalogErrorKind::EntryError(
                "Function definition has no name".into(),
            )]))
        }
    }

    fn add_type_entry(&mut self, entry: TypeEnvEntry<'_>) -> Result<ObjectId, CatalogError> {
        let eid = self
            .types
            .add(entry.name.as_ref(), entry.aliases.as_slice(), entry.ty);

        match eid {
            Ok(eid) => Ok((self.id, eid).into()),
            Err(e) => Err(e),
        }
    }

    fn get_function(&self, name: &str) -> Option<FunctionEntry<'_>> {
        self.functions
            .find_by_name(name)
            .map(|(e, f)| self.to_function_entry(e, f))
    }

    fn get_function_by_id(&self, id: ObjectId) -> Option<FunctionEntry<'_>> {
        assert_eq!(self.id, id.catalog_id());
        self.functions
            .find_by_id(&id.entry_id())
            .map(|(e, f)| self.to_function_entry(e, f))
    }

    fn resolve_type(&self, name: &str) -> Option<TypeEntry> {
        self.types.find_by_name(name).map(|(eid, entry)| TypeEntry {
            id: (self.id, eid).into(),
            ty: entry.clone(),
        })
    }
}

impl PartiqlCatalog {
    fn to_function_entry<'a>(
        &'a self,
        eid: EntryId,
        entry: &'a FunctionEntryFunction,
    ) -> FunctionEntry<'a> {
        FunctionEntry {
            id: (self.id, eid).into(),
            function: entry,
        }
    }
}

#[derive(Debug)]
struct CatalogEntrySet<T> {
    entries: HashMap<EntryId, T>,
    by_name: HashMap<UniCase<String>, EntryId>,
    by_alias: HashMap<UniCase<String>, EntryId>,

    next_id: AtomicU64,
}

impl<T> Default for CatalogEntrySet<T> {
    fn default() -> Self {
        CatalogEntrySet {
            entries: Default::default(),
            by_name: Default::default(),
            by_alias: Default::default(),
            next_id: 1.into(),
        }
    }
}

impl<T> CatalogEntrySet<T> {
    fn add(&mut self, name: &str, aliases: &[&str], info: T) -> Result<EntryId, CatalogError> {
        let mut errors = vec![];
        let name = UniCase::from(name);
        let aliases: Vec<UniCase<String>> = aliases
            .iter()
            .map(|a| UniCase::from((*a).to_string()))
            .collect();

        for a in &aliases {
            if self.by_alias.contains_key(a) {
                errors.push(CatalogErrorKind::EntryExists(a.as_ref().to_string()));
            }
        }

        if self.by_name.contains_key(&name) {
            errors.push(CatalogErrorKind::EntryExists(name.to_string()));
        }

        let id = self.next_id.fetch_add(1, Ordering::SeqCst).into();

        if let Some(_old_val) = self.entries.insert(id, info) {
            errors.push(CatalogErrorKind::Unknown);
        }

        match errors.is_empty() {
            true => {
                self.by_name.insert(name, id);

                for a in aliases {
                    self.by_alias.insert(a, id);
                }

                Ok(id)
            }
            _ => Err(CatalogError::new(errors)),
        }
    }

    fn find_by_id(&self, eid: &EntryId) -> Option<(EntryId, &T)> {
        self.entries.get(eid).map(|e| (*eid, e))
    }

    fn find_by_name(&self, name: &str) -> Option<(EntryId, &T)> {
        let name = UniCase::from(name);
        let eid = self.by_name.get(&name).or(self.by_alias.get(&name));

        eid.and_then(|eid| self.find_by_id(eid))
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn todo() {}
}
