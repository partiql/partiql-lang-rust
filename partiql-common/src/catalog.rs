#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct CatalogId(u64);

impl From<u64> for CatalogId {
    fn from(value: u64) -> Self {
        CatalogId(value)
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct EntryId(u64);

impl From<u64> for EntryId {
    fn from(value: u64) -> Self {
        EntryId(value)
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct ObjectId {
    catalog_id: CatalogId,
    entry_id: EntryId,
}

impl ObjectId {
    pub fn new(catalog_id: CatalogId, entry_id: EntryId) -> Self {
        Self {
            catalog_id,
            entry_id,
        }
    }

    pub fn catalog_id(&self) -> CatalogId {
        self.catalog_id
    }
    pub fn entry_id(&self) -> EntryId {
        self.entry_id
    }
}

impl From<(CatalogId, EntryId)> for ObjectId {
    fn from((catalog_id, entry_id): (CatalogId, EntryId)) -> Self {
        ObjectId::new(catalog_id, entry_id)
    }
}
