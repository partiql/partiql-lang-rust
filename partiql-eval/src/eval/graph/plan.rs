use std::fmt::Debug;
use std::hash::Hash;

use crate::eval::graph::types::{GraphElement, GraphTypes};
use fxhash::FxBuildHasher;
use indexmap::IndexSet;

type FxIndexSet<T> = IndexSet<T, FxBuildHasher>;

#[derive(Debug, Clone, Copy)]
pub enum DirSpec {
    L,   // <-
    U,   //  ~
    R,   //  ->
    LU,  // <~
    UR,  //  ~>
    LR,  // <->
    LUR, //  -
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct BindSpec<GT: GraphTypes>(pub GT::Binder);

#[derive(Debug, Clone, Default)]
pub enum LabelSpec<GT: GraphTypes> {
    #[default]
    Always,
    Named(GT::Label),
    Never,
}

#[derive(Debug, Clone, Copy, Default)]
pub enum FilterSpec {
    #[default]
    Always,
}

#[derive(Debug, Clone)]
pub struct NodeSpec<GT: GraphTypes> {
    pub label: LabelSpec<GT>,
    pub filter: FilterSpec,
}

pub trait ElemSpecBuilder<GT: GraphTypes> {
    fn any() -> Self;
    fn any_labeled(label: GT::Label) -> Self;
}

impl<GT: GraphTypes> ElemSpecBuilder<GT> for NodeSpec<GT> {
    fn any() -> Self {
        Self {
            label: LabelSpec::Always,
            filter: FilterSpec::Always,
        }
    }

    fn any_labeled(label: GT::Label) -> Self {
        Self {
            label: LabelSpec::Named(label),
            filter: FilterSpec::Always,
        }
    }
}

impl<GT: GraphTypes> ElemSpecBuilder<GT> for EdgeSpec<GT> {
    fn any() -> Self {
        Self {
            label: LabelSpec::Always,
            filter: FilterSpec::Always,
        }
    }
    fn any_labeled(label: GT::Label) -> Self {
        Self {
            label: LabelSpec::Named(label),
            filter: FilterSpec::Always,
        }
    }
}

#[derive(Debug, Clone)]
pub struct EdgeSpec<GT: GraphTypes> {
    pub label: LabelSpec<GT>,
    pub filter: FilterSpec,
}

#[derive(Debug, Clone)]
pub struct TripleSpec<GT: GraphTypes> {
    pub lhs: NodeSpec<GT>,
    pub e: EdgeSpec<GT>,
    pub rhs: NodeSpec<GT>,
}

#[derive(Debug, Clone)]
pub struct StepSpec<GT: GraphTypes> {
    pub dir: DirSpec,
    pub triple: TripleSpec<GT>,
}

#[derive(Debug, Clone)]
pub struct NodeMatch<GT: GraphTypes> {
    pub binder: BindSpec<GT>,
    pub spec: NodeSpec<GT>,
}

#[derive(Debug, Clone)]
pub struct EdgeMatch<GT: GraphTypes> {
    pub binder: BindSpec<GT>,
    pub spec: EdgeSpec<GT>,
}

#[derive(Debug, Clone)]
pub struct PathMatch<GT: GraphTypes> {
    pub binders: (BindSpec<GT>, BindSpec<GT>, BindSpec<GT>),
    pub spec: StepSpec<GT>,
}

#[derive(Debug, Clone)]
pub struct PathBinding<GT: GraphTypes> {
    pub matcher: PathMatch<GT>,
    pub bindings: Vec<Triple<GT>>,
}

#[derive(Debug, Clone)]
pub struct NodeBinding<GT: GraphTypes> {
    pub matcher: NodeMatch<GT>,
    pub binding: Vec<GT::NodeId>,
}

#[derive(Debug, Clone)]
pub struct Triple<GT: GraphTypes> {
    pub lhs: GT::NodeId,
    pub e: GT::EdgeId,
    pub rhs: GT::NodeId,
}

#[derive(Debug, Clone)]
pub enum PathPatternMatch<GT: GraphTypes> {
    Node(NodeMatch<GT>),
    Match(PathMatch<GT>),
    Concat(Vec<PathPatternMatch<GT>>),
    // Alternative(Vec<PathPatternMatch<GT>>), ?
}

impl<GT: GraphTypes> From<PathMatch<GT>> for PathPatternMatch<GT> {
    fn from(value: PathMatch<GT>) -> Self {
        Self::Match(value)
    }
}

impl<GT: GraphTypes> From<NodeMatch<GT>> for PathPatternMatch<GT> {
    fn from(value: NodeMatch<GT>) -> Self {
        Self::Node(value)
    }
}

#[derive(Debug, Clone)]
pub struct PathPatternSpecs<GT: GraphTypes> {
    pub head: NodeSpec<GT>,
    pub tail: Vec<(DirSpec, EdgeSpec<GT>, NodeSpec<GT>)>,
}

#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathPatternNodes<GT: GraphTypes> {
    pub head: GT::NodeId,
    pub tail: Vec<(GT::EdgeId, GT::NodeId)>,
}

impl<GT: GraphTypes> PathPatternNodes<GT> {
    pub fn len(&self) -> usize {
        1 + (self.tail.len() * 2)
    }

    pub fn is_empty(&self) -> bool {
        false
    }

    pub fn iter(&self) -> impl Iterator<Item = GraphElement<'_, GT>> {
        let head = std::iter::once(GraphElement::Node(&self.head));
        let tail = self
            .tail
            .iter()
            .flat_map(|(e, n)| [GraphElement::Edge(e), GraphElement::Node(n)].into_iter());
        head.chain(tail)
    }
}

#[derive(Debug, Clone)]
pub struct PathPatternBinding<GT: GraphTypes> {
    pub binders: Vec<BindSpec<GT>>,
    // pub specs: Vec<PathPatternSpecs<GT>>,
    pub bindings: FxIndexSet<PathPatternNodes<GT>>,
}

impl<GT: GraphTypes> From<PathBinding<GT>> for PathPatternBinding<GT> {
    fn from(path_binding: PathBinding<GT>) -> Self {
        let PathBinding { bindings, matcher } = path_binding;

        let (l, e, r) = matcher.binders;
        let binders = vec![l, e, r];
        let bindings = bindings
            .into_iter()
            .map(|triple| PathPatternNodes {
                head: triple.lhs,
                tail: vec![(triple.e, triple.rhs)],
            })
            .collect();

        PathPatternBinding { binders, bindings }
    }
}

impl<GT: GraphTypes> From<NodeBinding<GT>> for PathPatternBinding<GT> {
    fn from(node_binding: NodeBinding<GT>) -> Self {
        let NodeBinding { binding, matcher } = node_binding;

        let binders = vec![matcher.binder];

        let bindings = binding
            .into_iter()
            .map(|node| PathPatternNodes {
                head: node,
                tail: vec![],
            })
            .collect();

        PathPatternBinding { binders, bindings }
    }
}
