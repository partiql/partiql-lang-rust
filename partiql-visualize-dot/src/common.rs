pub trait ToDotGraph<T> {
    fn to_graph(self, data: &T) -> String;
}

pub(crate) const FG_COLOR: &'static str = "\"#839496\"";
