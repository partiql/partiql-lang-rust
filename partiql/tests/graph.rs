use insta::assert_snapshot;
use partiql_catalog::catalog::PartiqlCatalog;
use partiql_eval::plan::EvaluationMode;
use partiql_extension_ion::decode::{IonDecodeResult, IonDecoderBuilder, IonDecoderConfig};

use crate::common::{compile, evaluate, lower, parse};
use partiql_extension_ion::Encoding;
use partiql_value::{tuple, Value};

mod common;

#[track_caller]
fn decode_ion_text(contents: &str, encoding: Encoding) -> IonDecodeResult {
    let reader = ion_rs_old::ReaderBuilder::new().build(contents)?;
    let mut iter =
        IonDecoderBuilder::new(IonDecoderConfig::default().with_mode(encoding)).build(reader)?;

    let val = iter.next();

    val.unwrap()
}

#[track_caller]
fn pass(contents: &str) {
    let result = decode_ion_text(contents, Encoding::PartiqlEncodedAsIon);
    assert!(result.is_ok());
    dbg!(&result.unwrap());
}

#[track_caller]
fn fail(contents: &str) {
    let result = decode_ion_text(contents, Encoding::PartiqlEncodedAsIon);
    dbg!(&result);
    assert!(result.is_err());
}

#[test]
fn rfc_0025() {
    pass(include_str!("resources/rfc0025-example.ion"))
}

#[test]
fn gpml_paper() {
    pass(include_str!("resources/gpml-paper-example.ion"))
}

#[test]
fn empty_graph() {
    pass("$graph::{ nodes: [], edges: [] }")
}
#[test]
fn graph_wo_annotations() {
    pass("$graph::{ nodes: [], edges: [] }")
}
#[test]
fn empty_graph_no_edges() {
    fail("$graph::{ nodes: [] }")
}

#[test]
fn node_with_id_label_payload() {
    pass(r##"$graph::{ nodes: [ { id: n1, labels: ["a"], payload: 33 } ], edges: [] }"##)
}
#[test]
fn node_with_labels() {
    pass(r##"$graph::{ nodes: [ { id: n1, labels: ["a", "b", "c"] } ], edges: [] }"##)
}
#[test]
fn node_no_labels() {
    pass(r##"$graph::{ nodes: [ { id: n1, labels: [] } ], edges: [] }"##)
}
#[test]
fn node_complex_payload() {
    pass(r##"$graph::{ nodes: [ { id: n1, payload: {height: 12, weight: 152} } ], edges: [] }"##)
}
#[test]
fn node_no_id() {
    fail(r##"$graph::{ nodes: [ { labels: ["a"], payload: "must have the id!" } ], edges: [] }"##)
}

#[test]
fn edge_with_label() {
    pass(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                              edges: [ {id: e1, labels: ["go"], ends: (n1 -> n2)} ] }"##,
    )
}
#[test]
fn edge_with_labels() {
    pass(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                               edges: [ {id: e1, labels: ["go", "went", "gone"], ends: (n1 -> n2)} ] }"##,
    )
}
#[test]
fn undirected_edge_no_labels() {
    pass(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                               edges: [ {id: e1, labels: [], ends: (n1 -- n2)} ] }"##,
    )
}
#[test]
fn undirected_edge_complex_payload() {
    pass(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                               edges: [ {id: e1, ends: (n1 -- n2), payload: {length: 23, thickness: 3} }] }"##,
    )
}
#[test]
fn node_to_itself() {
    pass(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                               edges: [ {id: e1, ends: (n1 -- n1)},
                                        {id: e2, ends: (n2 <- n2)}, ] }"##,
    )
}
#[test]
fn edge_no_id() {
    fail(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                               edges: [ {labels: ["a"], ends: (n2 <- n1)} ] }"##,
    )
}
#[test]
fn edge_no_ends() {
    fail(
        r##"$graph::{ nodes: [{id: n1}, {id: n2}],
                               edges: [ { id: e1, labels: ["a"] } ] }"##,
    )
}

#[test]
fn nodes2_edges3() {
    pass(
        r##"$graph::{
                        nodes: [ {id: n1}, {id: n2} ],
                        edges: [ {id: e1, ends: (n1 -> n2)},
                                 {id: e2, ends: (n1 <- n2)},
                                 {id: e3, ends: (n1 -- n2)} ] }"##,
    )
}

#[test]
fn shared_identifiers() {
    pass(r##"$graph::{nodes: [ {id: x}, {id: y} ], edges: [ {id:x, ends: (x -> y)} ] }"##)
}

// The following examples are valid per ISL, but should be rejected by a processor.
#[test]
#[ignore] // TODO: this currently panics; fix to return error
fn edges_bad_nodes() {
    fail(r##"$graph::{nodes: [ {id: n1}, {id: n2} ], edges: [ {id:e1, ends: (z1 -> z2)} ] }"##)
}
#[test]
#[ignore] // TODO: this currently panics; fix to return error
fn repeated_identifers() {
    fail(
        r##"$graph::{
                        nodes: [ {id: n1}, {id: n1} ],
                        edges: [ {id:e1, ends: (n1 -> n1)},
                                 {id:e1, ends: (n1 -- n1)} ] }"##,
    )
}

#[track_caller]
fn snapshot_test_graph_eval(name: &'static str, contents: &'static str, query: &'static str) {
    let graph = decode_ion_text(contents, Encoding::PartiqlEncodedAsIon).expect("graph decode");
    let bindings = tuple![("g", graph)];

    let parsed = parse(query).expect("parse");
    let catalog = PartiqlCatalog::default();
    let lowered = lower(&catalog, &parsed).expect("lower");
    let plan = compile(EvaluationMode::Permissive, &catalog, lowered).expect("compile");
    let res = evaluate(plan, bindings.into());
    assert!(res.is_ok());
    assert_snapshot!(name, res.unwrap().result);
}

#[test]
fn select_star_rfc_0025() {
    // The RFC 0025 graph has no payloads so these just return an appropriate number of `{}`
    let contents = include_str!("resources/rfc0025-example.ion");
    snapshot_test_graph_eval("RFC0025 Nodes", contents, "(g MATCH (x))");
    snapshot_test_graph_eval("RFC0025 L Triples", contents, "(g MATCH (x) -> (y))");
    snapshot_test_graph_eval("RFC0025 LUR Triples", contents, "(g MATCH (x) - (y))");
}

#[test]
fn select_star_gpml_paper() {
    let contents = include_str!("resources/gpml-paper-example.ion");
    snapshot_test_graph_eval("GPML Nodes", contents, "(g MATCH (x))");
    snapshot_test_graph_eval("GPML L Triples", contents, "(g MATCH (x) -[e]-> (y))");
    snapshot_test_graph_eval("GPML LUR Triples", contents, "(g MATCH (x) -[e]- (y))");
}
