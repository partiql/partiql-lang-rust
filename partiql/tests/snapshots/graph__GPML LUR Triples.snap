---
source: partiql/tests/graph.rs
expression: res.unwrap().result
---
<<
  {
    'x': { 'owner': 'Scott', 'isBlocked': 'no' },
    'y': { 'number': '123.111', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '123.111', 'isBlocked': 'no' },
    'y': { 'owner': 'Scott', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '123.222', 'isBlocked': 'no' },
    'y': { 'owner': 'Charles', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Charles', 'isBlocked': 'no' },
    'y': { 'number': '123.222', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '111', 'isBlocked': 'no' },
    'y': { 'owner': 'Scott', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Scott', 'isBlocked': 'no' },
    'y': { 'number': '111', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '222', 'isBlocked': 'no' },
    'y': { 'owner': 'Aretha', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Aretha', 'isBlocked': 'no' },
    'y': { 'number': '222', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Mike', 'isBlocked': 'no' },
    'y': { 'number': '222', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '222', 'isBlocked': 'no' },
    'y': { 'owner': 'Mike', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Jay', 'isBlocked': 'yes' },
    'y': { 'number': '333', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '333', 'isBlocked': 'no' },
    'y': { 'owner': 'Jay', 'isBlocked': 'yes' }
  },
  {
    'x': { 'number': '111', 'isBlocked': 'no' },
    'y': { 'owner': 'Charles', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Charles', 'isBlocked': 'no' },
    'y': { 'number': '111', 'isBlocked': 'no' }
  },
  {
    'x': { 'number': '444', 'isBlocked': 'no' },
    'y': { 'owner': 'Dave', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Dave', 'isBlocked': 'no' },
    'y': { 'number': '444', 'isBlocked': 'no' }
  },
  { 'x': { 'owner': 'Scott', 'isBlocked': 'no' }, 'y': { 'name': 'Zembla' } },
  { 'x': { 'name': 'Zembla' }, 'y': { 'owner': 'Scott', 'isBlocked': 'no' } },
  {
    'x': { 'owner': 'Aretha', 'isBlocked': 'no' },
    'y': { 'name': 'Ankh-Morpork' }
  },
  {
    'x': { 'name': 'Ankh-Morpork' },
    'y': { 'owner': 'Aretha', 'isBlocked': 'no' }
  },
  { 'x': { 'owner': 'Mike', 'isBlocked': 'no' }, 'y': { 'name': 'Zembla' } },
  { 'x': { 'name': 'Zembla' }, 'y': { 'owner': 'Mike', 'isBlocked': 'no' } },
  {
    'x': { 'owner': 'Jay', 'isBlocked': 'yes' },
    'y': { 'name': 'Ankh-Morpork' }
  },
  {
    'x': { 'name': 'Ankh-Morpork' },
    'y': { 'owner': 'Jay', 'isBlocked': 'yes' }
  },
  { 'x': { 'owner': 'Charles', 'isBlocked': 'no' }, 'y': { 'name': 'Zembla' } },
  { 'x': { 'name': 'Zembla' }, 'y': { 'owner': 'Charles', 'isBlocked': 'no' } },
  {
    'x': { 'owner': 'Scott', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-01-01 0:00:00.0', 'amount': 8000000.00 },
    'y': { 'owner': 'Mike', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Mike', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-01-01 0:00:00.0', 'amount': 8000000.00 },
    'y': { 'owner': 'Scott', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Mike', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-02-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Aretha', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Aretha', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-02-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Mike', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Aretha', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-03-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Jay', 'isBlocked': 'yes' }
  },
  {
    'x': { 'owner': 'Jay', 'isBlocked': 'yes' },
    'e': { 'date': TIMESTAMP '2020-03-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Aretha', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Jay', 'isBlocked': 'yes' },
    'e': { 'date': TIMESTAMP '2020-04-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Dave', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Dave', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-04-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Jay', 'isBlocked': 'yes' }
  },
  {
    'x': { 'owner': 'Dave', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-06-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Mike', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Mike', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-06-01 0:00:00.0', 'amount': 10000000.00 },
    'y': { 'owner': 'Dave', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Dave', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-07-01 0:00:00.0', 'amount': 4000000.00 },
    'y': { 'owner': 'Charles', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Charles', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-07-01 0:00:00.0', 'amount': 4000000.00 },
    'y': { 'owner': 'Dave', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Mike', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-08-01 0:00:00.0', 'amount': 6000000.00 },
    'y': { 'owner': 'Charles', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Charles', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-08-01 0:00:00.0', 'amount': 6000000.00 },
    'y': { 'owner': 'Mike', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Charles', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-09-01 0:00:00.0', 'amount': 9000000.00 },
    'y': { 'owner': 'Scott', 'isBlocked': 'no' }
  },
  {
    'x': { 'owner': 'Scott', 'isBlocked': 'no' },
    'e': { 'date': TIMESTAMP '2020-09-01 0:00:00.0', 'amount': 9000000.00 },
    'y': { 'owner': 'Charles', 'isBlocked': 'no' }
  }
>>
